habitat/core/vector_env.py

```diff
@@ -75,6 +76,30 @@ def _make_env_fn(
     return habitat_env
 
 
+@attr.s(auto_attribs=True, slots=True)
+class ReadWrapper:
+    read_fn: Callable[[], Any]
+    _is_waiting: bool = False
+
+    def __call__(self) -> Any:
+        assert self._is_waiting
+        res = self.read_fn()
+        self._is_waiting = False
+
+        return res
+
+
+@attr.s(auto_attribs=True, slots=True)
+class WriteWrapper:
+    write_fn: Callable[[Any], None]
+    _read_wrapper: ReadWrapper
+
+    def __call__(self, data: Any) -> None:
+        assert not self._read_wrapper._is_waiting
+        self.write_fn(data)
+        self._read_wrapper._is_waiting = True
+
+
 class VectorEnv:
```

Comment:
As an aside, shouldn't VectorEnv by sub-classing the Gym VectorEnv?

Reply:
Or at the very leasr Habitat.Env or something?

Reply:
Seems odd to have the vector version inherit from the single env version given how different the APIs are.  I don't think the vector env passing `isinstance(env, habitat.Env)` makes sense.

Reply:
Fair enough, but it should inherit from the GymVectorEnv, right?

Reply:
We have a slightly different API, we return a list of each environment's output instead of `obs, rewards, dones, info` to not deeply enforce that pattern (same reason why `habitat.Env` does not inherit from `gym.Env`).

Reply:
Correct, Habitat Lab core was designed to be open to different approaches. That's the reason we don't provide reward as part of the task specification.